### ISSUE TITLE
click sound when track 1 stops recording

### ISSUE THREAD

**Issue by ciscokid009** [RandomVertebrate/raspi-looper#6]
this applies only to track 1, it does not occur on tracks 2-4.
can't really say if it's a record or playback issue.

**Comment by RandomVertebrate** [comment]
I don't know what this is. Could be distortion coincidence. Would
appreciate if you send me a recording.

**Comment by ciscokid009** [comment]
tried again with microphone, no click sound.
tried with guitar - click sound.

i suspect the click-sound occurs when i plug in my guitar, play, and there is still guitar sound when i stop recording. i noticed similiar clicking sounds when i did wave-file editing in audacity, for example.

i think the clicking occurs because by the end of the recorded audio-file there is wave data with volume very much higher than 0, but the loop starts with 0 bias, so when the loop starts again, there is a volume decrease from last value from guitar sound to zero after only one time tick (1s/44100).

same problem will be when starting a recording with already guitar sound, and stopping with no sound.

workaround could be be a kind of fade-out to zero bias within a couple of milliseconds when stopping the recording, and fading playback same way before looping.

**Comment by ciscokid009** [comment]
i think this will also address the clipping/mixing issue.

**Comment by RandomVertebrate** [comment]
Does this always happen only at exactly the beginning of the loop?

**Comment by RandomVertebrate** [comment]
> i suspect the click-sound occurs when i plug in my guitar, play, and there is still guitar sound when i stop recording. i noticed similiar clicking sounds when i did wave-file editing in audacity, for example.

A recording would really help me grasp what's going on here

**Comment by ciscokid009** [comment]
sorry for the delay, is there an easy way to save the recorded wave within main.py, maybe with an additional switch on another gpio? i do have a another one mounted, not in use yet. 

i managed to eliminate distortions originated by the dc-power supply (the cheap soundcard i am using obviously doesn't filter out anything) by using a power bank battery for smartphones instead, but i only have a terrible mic, where it's pretty useless to record the recording again, the click sound isn't really noticable, also with my smartphone a recording isn't really usefull.

**Comment by RandomVertebrate** [comment]
I think this is happening during mixing. It could in fact be due to a bias.

  here, this version of main.py should write the loops to files locally on
the raspberry pi sd card

**Comment by ciscokid009** [comment]
sorry, i don't see any link to the new main.py

**Comment by RandomVertebrate** [comment]
Sorry about that. Here it is:
https://drive.google.com/file/d/1yup7ZvmRYKR9sxXXjtEui6w6-yp7NPIP/view?usp=sharing

**Comment by ciscokid009** [comment]
thanks for the main.py file, as expected, there is no click sound recorded:

https://ufile.io/antb0afw

when you open it in audacity you will see that the wave ends far above a bias of zero.
playing it in audacity brings up exactly the same click sound when the playing stops.

![grafik](https://user-images.githubusercontent.com/85055236/121772599-91f0f800-cb76-11eb-980f-c48e74051be0.png)

**Comment by RandomVertebrate** [comment]
This is very interesting! What's more your fade-out, fade-in solution seems
to work in Audacity :)
I shall implement this into the python script soon.

**Comment by RandomVertebrate** [comment]
To be clear, this only happens on track 1, right? And only on the first recording, not on the overdubs? because implementing it on overdubs would be much harder :P

**Comment by ciscokid009** [comment]
i will test on all tracks and overdubs, if i get some spare time today ;)
but i expect it to happen on all tracks, it's physics, the membrane of the speaker get's pushed hard from zero bias to something hearable within 1/44100 second.

**Comment by RandomVertebrate** [comment]
I don't really understand why it's happening; your explanation of the physics doesn't really align with the way I understand signal processing or vibration theory. I think maybe a more intuitive way to look at this might be in the frequency domain in terms of waterfall graphs.

Anyway, the fact remains that it is happening. I'll send you a modified script which adds fades to the beginning and end of the very first recording of track 1 but nowhere else. Then you can let me know where else the clicks occur.

**Comment by ciscokid009** [comment]
great, looking forward!

**Comment by RandomVertebrate** [comment]
Hey, so I tried fade-outs and fade-ins but that doesn't really work in a looping context because you lose a bit of the beginning and end of your loop. So I added some code to keep track of the previously recorded buffer as an extra bit of audio at the beginning and made a cross-fade with that at the end. I also changed some stuff that was stopping the master track from recording again if it is cleared/deleted. I think that should work now though I'm not sure and I probably wouldn't do that. let me know if your issue seems sorted; if so I'll update the repository with this version.
https://drive.google.com/file/d/1YG_ADkp9vNCzgUf8BO1k8xKl17ZPeqIR/view?usp=sharing
